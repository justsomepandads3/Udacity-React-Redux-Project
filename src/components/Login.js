import { useEffect, useState } from "react";
import image from "../utils/LoginPageImage.svg"
import { connect } from "react-redux";
import { _getUsers } from "../utils/_DATA";
import { setAuthedUser } from "../actions/authedUser";
import { useNavigate } from "react-router-dom";
const Login = (props) =>{
    
    const nav=useNavigate(props.page)
    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    
    const users = Object.values(props.users);

    //The error part is inspired by code with yourself https://www.youtube.com/watch?v=S6pp_bpgMJg
    const [errors,setErrors] = useState({});
  

    const loginCred = () =>{
       
        users.forEach(user =>{
            if(user.id===username && user.password===password){
                props.dispatch(setAuthedUser(user.id))
                console.log(props.path)
                if(props.path==="/"){
                    nav("/dashboard")
                    console.log(props.path)
                }

                else{ 
                    nav(props.path)
                }

                
                
            }else{
                setErrors(validate(user.id===username,user.password===password))
            }
            
        })
        setErrors(validate())
    }
    
    //The error part is inspired by code with yourself https://www.youtube.com/watch?v=S6pp_bpgMJg
    const validate = (user=true, pass=true) =>{
       
        const errors = {}
        if(!username)
            errors.username="Enter a username";
        else if(/\S+@\S+\./.test(username))
            errors.username="Enter a valid username";
        else if(user && pass)
            errors.username="User or password is wrong"
        else 
            errors.username = "";
        if(!password)
            errors.password="Enter a password";
        else if(/\S+@\S+\./.test(password))
            errors.password="Enter a valid password";
        else
            errors.password = "";
        return errors
    }

    const updatePassword = (value) =>{
        setPassword(value);
    }

    const updateUsername = (value) =>{
        setUsername(value);
    }
    
    const errorText= (error) =>{
            return <div className="error" id="error-text" data-testid="error-text">{error}</div>
    }

    return(
        <div className="login">
            <img src={image} alt="img"></img>
            <h1>Login</h1>
            <h4>User</h4>
            <form >
                <input 
                    id="username-input"
                    data-testid="username-input"
                    type="text"
                    placeholder="Enter you Username" 
                    value={username}
                    onChange={e => updateUsername(e.target.value)}
                    
                ></input>
                {errors.username && errorText(errors.username)}
                <h4>Password</h4>
                <input 
                    id="password-input"
                    data-testid="password-input"
                    type ="Password" 
                    placeholder="Enter your Password"
                    value={password}
                    onChange={e => updatePassword(e.target.value)} 
                    
                ></input>
                {errors.password && errorText(errors.password)}

                <div className="buttonDiv"> <button id="login-submit-button" data-testid="login-submit-button" className="button" type="button" onClick={loginCred}>Log in</button></div>
                
            </form>
        </div>
    )
}
const mapStateToProps = ({ users,questions,authedUser }) => {
    return{
    users,
    questions,
    authedUser,
    
}};
export default connect(mapStateToProps)(Login);