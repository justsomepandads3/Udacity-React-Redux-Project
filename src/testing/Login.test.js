import { fireEvent, getByTestId, render } from "@testing-library/react";
import Login from "../components/Login";
import * as React from "react"
import reducer from "../reducers";
import middlewares from "../middleware";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";

const store = createStore(reducer, middlewares)

describe("Login component", () =>{
    it('will match the snapshot',()=>{
        var component = render(
        <Provider store={store}>
            <BrowserRouter>
                <Login></Login>
            </BrowserRouter>

        </Provider>
    );
        expect(component).toMatchSnapshot();
    })
    it("Logged in successfully" ,() =>{
        var component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login></Login>
                </BrowserRouter>
    
            </Provider>
        )
        var usernameInput = component.getByTestId("username-input")
        var passwordInput = component.getByTestId("password-input")
        var submitButton = component.getByTestId("login-submit-button")

        expect(usernameInput).toBeInTheDocument;
        expect(passwordInput).toBeInTheDocument;
        expect(submitButton).toBeInTheDocument;

        fireEvent.change(usernameInput, {target: {value: 'mtsamis'}})
        fireEvent.change(passwordInput, {target: {value: 'xyz123'}})
        fireEvent.click(submitButton)

        expect(component.getByTestId("error-text")).not.toBeInTheDocument;





    })

    it("The input fields are empty" ,() =>{
        var component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login></Login>
                </BrowserRouter>
    
            </Provider>
        )
        var usernameInput = component.getByTestId("username-input")
        var passwordInput = component.getByTestId("password-input")
        var submitButton = component.getByTestId("login-submit-button")

        expect(usernameInput).toBeInTheDocument;
        expect(passwordInput).toBeInTheDocument;
        expect(submitButton).toBeInTheDocument;

        fireEvent.change(usernameInput, {target: {value: ''}});
        fireEvent.change(passwordInput, {target: {value: ''}});
        fireEvent.click(submitButton);

        expect(component.getAllByTestId("error-text")).toBeInTheDocument;





    })
    it("Not logged in successfully" ,() =>{
        var component = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login></Login>
                </BrowserRouter>
    
            </Provider>
        )
        var usernameInput = component.getByTestId("username-input")
        var passwordInput = component.getByTestId("password-input")
        var submitButton = component.getByTestId("login-submit-button")

        expect(usernameInput).toBeInTheDocument;
        expect(passwordInput).toBeInTheDocument;
        expect(submitButton).toBeInTheDocument;

        fireEvent.change(usernameInput, {target: {value: 'mtmis'}});
        fireEvent.change(passwordInput, {target: {value: 'xy23'}});
        fireEvent.click(submitButton);

        expect(component.getByTestId("error-text")).toBeInTheDocument;





    })

})