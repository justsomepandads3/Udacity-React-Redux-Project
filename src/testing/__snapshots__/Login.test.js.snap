// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component will match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login"
      >
        <img
          alt="img"
          src="LoginPageImage.svg"
        />
        <h1>
          Login
        </h1>
        <h4>
          User
        </h4>
        <form>
          <input
            data-testid="username-input"
            id="username-input"
            placeholder="Enter you Username"
            type="text"
            value=""
          />
          <h4>
            Password
          </h4>
          <input
            data-testid="password-input"
            id="password-input"
            placeholder="Enter your Password"
            type="Password"
            value=""
          />
          <div
            class="buttonDiv"
          >
             
            <button
              class="button"
              data-testid="login-submit-button"
              id="login-submit-button"
              type="button"
            >
              Log in
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login"
    >
      <img
        alt="img"
        src="LoginPageImage.svg"
      />
      <h1>
        Login
      </h1>
      <h4>
        User
      </h4>
      <form>
        <input
          data-testid="username-input"
          id="username-input"
          placeholder="Enter you Username"
          type="text"
          value=""
        />
        <h4>
          Password
        </h4>
        <input
          data-testid="password-input"
          id="password-input"
          placeholder="Enter your Password"
          type="Password"
          value=""
        />
        <div
          class="buttonDiv"
        >
           
          <button
            class="button"
            data-testid="login-submit-button"
            id="login-submit-button"
            type="button"
          >
            Log in
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
